File: razer-sdk/src/test/java/com/bonepl/razersdk/effects/BlackToNoneEffectTest.java
package com.bonepl.razersdk.effects;

import com.bonepl.razersdk.ChromaRestSDK;
import com.bonepl.razersdk.animation.AnimatedFrame;
import com.bonepl.razersdk.animation.Animation;
import com.bonepl.razersdk.color.Color;
import com.bonepl.razersdk.animation.SimpleFrame;
import com.bonepl.razersdk.color.StaticColor;
import com.bonepl.razersdk.sdk.RzKey;
import org.junit.jupiter.api.Test;

import java.util.List;

import static com.bonepl.razersdk.sdk.RzKey.*;

public class BlackToNoneEffectTest {
    private static final List<RzKey> FIRST_LETTERS =
            List.of(RZKEY_Q, RZKEY_W, RZKEY_E, RZKEY_R, RZKEY_T, RZKEY_Y, RZKEY_U);

    @Test
    @SuppressWarnings("JUnitTestMethodWithNoAssertions")
    void testNonTransparentBlack() throws InterruptedException {
        try (ChromaRestSDK razerSDKClient = new ChromaRestSDK()) {
            Animation animation = new Animation();
            animation.addToFront(createMovingBlackPixel(StaticColor.BLACK));
            while (animation.hasFrame()) {
                animation.addToBack(new SimpleFrame(FIRST_LETTERS, StaticColor.CYAN));
                razerSDKClient.createKeyboardEffect(animation);
                Thread.sleep(100);
            }
        }
    }

    @Test
    @SuppressWarnings("JUnitTestMethodWithNoAssertions")
    void testTransparentBlack() throws InterruptedException {
        try (ChromaRestSDK razerSDKClient = new ChromaRestSDK()) {
            Animation animation = new Animation();
            animation.addToFront(createMovingBlackPixel(StaticColor.NONE));
            while (animation.hasFrame()) {
                animation.addToBack(new SimpleFrame(FIRST_LETTERS, StaticColor.CYAN));
                razerSDKClient.createKeyboardEffect(animation);
                Thread.sleep(100);
            }
        }
    }

    private static AnimatedFrame createMovingBlackPixel(Color color) {
        AnimatedFrame movingBlackPixel = new AnimatedFrame();
        FIRST_LETTERS.forEach(rzKey -> movingBlackPixel.addAnimationFrame(2, new SimpleFrame(rzKey, color)));
        return movingBlackPixel;
    }
}

File: test/razer-sdk/effects/BlackToNoneEffect.test.ts
import {describe, test} from 'vitest';
import {ChromaNativeSDK} from '../../../src/razer-sdk/ChromaNativeSDK.js';
import {AnimatedFrame} from '../../../src/razer-sdk/animation/AnimatedFrame.js';
import {Animation} from '../../../src/razer-sdk/animation/Animation.js';
import {SimpleFrame} from '../../../src/razer-sdk/animation/SimpleFrame.js';
import {StaticColor} from '../../../src/razer-sdk/color/StaticColor.js';
import {Color} from '../../../src/razer-sdk/color/Color.js';
import {RzKey} from '../../../src/razer-sdk/sdk/RzKey.js';

const FIRST_LETTERS: RzKey[] = [
    RzKey.RZKEY_Q, RzKey.RZKEY_W, RzKey.RZKEY_E, RzKey.RZKEY_R,
    RzKey.RZKEY_T, RzKey.RZKEY_Y, RzKey.RZKEY_U,
];

function createMovingBlackPixel(color: Color): AnimatedFrame {
    const movingBlackPixel = new AnimatedFrame();
    for (const rzKey of FIRST_LETTERS) {
        movingBlackPixel.addAnimationFrame(2, new SimpleFrame(rzKey, color));
    }
    return movingBlackPixel;
}

/**
 * Integration test comparing opaque black vs transparent black on actual hardware.
 * Demonstrates layering behavior with BLACK (overwrites) vs NONE (transparent).
 */
describe('BlackToNoneEffect Integration', () => {
    test('testNonTransparentBlack', async () => {
        const sdk = new ChromaNativeSDK();
        await sdk.connect();

        try {
            const animation = new Animation();
            animation.addToFront(createMovingBlackPixel(StaticColor.BLACK));

            while (animation.hasFrame()) {
                animation.addToBack(new SimpleFrame(FIRST_LETTERS, StaticColor.CYAN));
                await sdk.createKeyboardEffect(animation);
                await new Promise(resolve => setTimeout(resolve, 100));
            }
        } finally {
            await sdk.close();
        }
    }, 30000);

    test('testTransparentBlack', async () => {
        const sdk = new ChromaNativeSDK();
        await sdk.connect();

        try {
            const animation = new Animation();
            animation.addToFront(createMovingBlackPixel(StaticColor.NONE));

            while (animation.hasFrame()) {
                animation.addToBack(new SimpleFrame(FIRST_LETTERS, StaticColor.CYAN));
                await sdk.createKeyboardEffect(animation);
                await new Promise(resolve => setTimeout(resolve, 100));
            }
        } finally {
            await sdk.close();
        }
    }, 30000);
});
